### ISSUE TITLE
How can I sleep 600 ms

### ISSUE THREAD

**Issue by Jean-Luc-Picard-2021** [trealla-prolog/trealla#134]
Hi, I tried:
```
/* Trealla Prolog 2.11.17 */
?- sleep(0.6).
   throw(error(type_error(integer,0.6),sleep/1)).
```
In SWI-Prolog I can do:
```
/* SWI-Prolog 9.1.4 */
?- sleep(0.6).
true.
```

**Comment by Jean-Luc-Picard-2021** [comment]
Also sleep/1 seems to be quite costly. Not that I wouldn't want
that sleep(0) does something and possibly yields the thread. 

But it has a high overhead:
```
?- time(fib(29,_)).
   % Time elapsed 0.716444s
   true.
?- time(fib_hell(29,_)).
   % Time elapsed 63.235085s
   true.
```
Was using this code:
```
fib(0, R) :- !, R=1.
fib(1, R) :- !, R=1.
fib(N, R) :- M is N-1, fib(M, A), L is M-1, fib(L, B), R is A+B.

fib_hell(0, R) :- !, R=1, sleep(0).
fib_hell(1, R) :- !, R=1, sleep(0).
fib_hell(N, R) :- M is N-1, fib_hell(M, A), L is M-1, fib_hell(L, B), R is A+B.
```

**Comment by infradig** [comment]
delay(600)

**Comment by infradig** [comment]
The cost of sleep(0) is the price you pay (SWI does the same(. There is
always the option not to call if the interval is zero.

**Comment by Jean-Luc-Picard-2021** [comment]
If `sleep(0)` would simply skip, the price would be very low. I have
two implementations of sleep/1. One that does a skip when delay=<0,
and one that does always a system call. Here is a comparison:
```
/* Jekejeke Prolog 1.5.6 */
/* sleep/1 that skips, from user module */
?- time(fib(29,_)).
% Up 435 ms, GC 3 ms (Current 03/05/23 02:33:44)
true.
?- time(fib_hell(29,_)).
% Up 466 ms, GC 2 ms (Current 03/05/23 02:33:46)
true.

/* Jekejeke Prolog 1.5.6 with debugger, slightly slower */
/* thread_sleep/1 that does always a system call, from module system/thread */
?- time(fib(29,_)).
% Up 457 ms, GC 2 ms (Current 03/05/23 02:22:03)
true.
?- time(fib_hell(29,_)).
% Up 650 ms, GC 2 ms (Current 03/05/23 02:22:06)
true.
```
So the foreign function call of sleep/1 and the skip delay=<0 makes
only ca. 10%. On the other hand the system call makes around 150% 
overhead.  How does it JDK 19 do? I don't know. What does it do?

**Comment by Jean-Luc-Picard-2021** [comment]
You can test the sleep/1 call overhead, if it would
do nothing by for example using this mockup:
```
thread_sleep(_).
```
I then get:
```
/* Trealla Prolog 2.11.17 */
/* mockup comparison */
?- time(fib(29,_)).
   % Time elapsed 0.619202s
   true.
?- time(fib_hell(29,_)).
   % Time elapsed 0.706442s
   true.
```
So its not that the sleep call fundamentally destroys fibonacci.
Looks like its rather a test case, which can measure the sleep/1 
calls, without causing disruption in the fibonacci function itself.

Your mockup overhead is roughly the same as my sleep/1 that skips.

**Comment by infradig** [comment]
I understand the sleep(0) overhead, but that's how it is.

**Comment by Jean-Luc-Picard-2021** [comment]
How can I use this here:
```
       if (q->is_task)
		return do_yield(q, get_smallint(p1));
```
What is the overhead of this sleep?

**Comment by Jean-Luc-Picard-2021** [comment]
How can I use this here:
```
       if (q->is_task)
		return do_yield(q, get_smallint(p1));
```
What is the overhead of this sleep?

Ok, I tried this. But I am not sure what time/1 is supposed
to do and measure as a task? Can it do that?
```
?- tasklist(time, [fib(29,_), fib_hell(29,_)]).
   % Time elapsed 0.781001s
%%% hangs
```
Somehow the yield sleep is much slower?

**Comment by infradig** [comment]
Tasks are a form of cooperative, concurrent processes. The scheduler at
this stage is very primitive and based around millsec waits when there is
no work scheduled to be done, and works best with I/O bound workloads (like
http gets). Something computationally heavy like FIB will hit a (brick)
wall trying to get work done. One day day I will upgrade it with event
based waits using select or similar.

**Comment by Jean-Luc-Picard-2021** [comment]
Do you have an engine abstraction in Trealla Prolog?
Like this here from SWI-Prolog:


> 11.3 Engine predicate reference
> https://www.swi-prolog.org/pldoc/man?section=engine-predicates

So everybody can write an event loop in Prolog itself?
Well the API is not perfect, a lot is missing right now.

**Comment by guregu** [comment]
You can implement your own event loop with Trealla's C API. That is what trealla-js and trealla-prolog/go do. The closest thing to engines would be `pl_create` and `pl_query`. Within Prolog, you can use `yield/0` to pause the query and return control. 
Here's an example of how trealla-js specializes sleep/0 by calling into JS: https://github.com/guregu/trealla/blob/02e188c7420016c7f63829af3450225a61fcebbc/library/wasm_js.pl#L128-L141
This is implemented by a special function on the host that associates a JS Promise with a Trealla subquery: 
https://github.com/guregu/trealla/blob/02e188c7420016c7f63829af3450225a61fcebbc/library/wasm_js.pl#L13-L15
https://github.com/guregu/trealla-js/blob/8b8b9fca619662604957d4f22dd7354ffb087915/src/prolog.js#L240
And in the query loop it checks if the query yielded and has a Promise it needs to await: https://github.com/guregu/trealla-js/blob/8b8b9fca619662604957d4f22dd7354ffb087915/src/prolog.js#L129-L139

Of course it's not the only way we could implement this, just the simplest way to get HTTP fetch and sleep working. You could also make a system like Tau's that binds a variable to an opaque Promise handle and some predicates to await it.

For trealla-go it's simpler because each interpreter runs in its own coroutine, but it has a mutex on the interpreter (i.e. engine) so multiple subqueries could get blocked by forever if the currently-running query hogs the thread. I'd like to investigate some auto-yielding thing like SWI's WASM implementation and Trealla's task stuff.

The biggest difference between SWI and Trealla's 'engine' concept is that it's single-threaded and doesn't have transactions or any of that, so it's up to you to synchronize the threads if you want to run them truly simultaneously. JS only has one thread so it's fine, for Go there's a mutex. I think we could probably do some stuff like copy-on-write and run multiple queries simultaneously if we hacked on it. You could do some crazy stuff with WASM like keep copies of the whole process memory and time-travel backwards on a conflict, it's definitely a cool thing to thing about.

Note that this stuff uses some unstable things from my experimental fork (like query_did_yield in trealla.h), but I'm interested in upstreaming it if it's useful.

**Comment by guregu** [comment]
FWIW, I like that Trealla is single-threaded, it make it easy to embed/port and you can leave the scheduling to the host's runtime. For example, you could call `pl_create` N times and run them on N threads just fine, but they won't share a knowledgebase. This should work with Web Workers on trealla-js too (each `new Prolog()` is a completely separate instance). It'd be cool to be able to share the DB somehow too, of course.

**Comment by Jean-Luc-Picard-2021** [comment]
Did you try this test case:
```
?- time(fib(29,_)).
   % Time elapsed 0.716444s
   true.
?- time(fib_hell(29,_)).
   % Time elapsed 63.235085s
   true.
```
With the phrase/2 and js_eval_/3 approach? 

Who even invents calling mechanisms of JavaScript that send
their code first to the north pole and then back, before
something happens on the machine?

**Comment by infradig** [comment]
> Do you have an engine abstraction in Trealla Prolog? Like this here from SWI-Prolog:
> 
> > 11.3 Engine predicate reference
> > https://www.swi-prolog.org/pldoc/man?section=engine-predicates
> 
> So everybody can write an event loop in Prolog itself? Well the API is not perfect, a lot is missing right now.

Engines should be doable. They share access to the database.

**Comment by guregu** [comment]
My point is not "you should implement sleep by dynamically calling JS",
it's "here's a minimal example of how it's possible to integrate with
another runtime's scheduler". The system I described is intended to be used
for awaiting HTTP fetches without blocking the main thread in the browser
version of Trealla, and it does a decent job of that. Try running multiple
`http_fetch("/", X, []).` queries concurrently here
https://php.energy/trealla.html and you'll see that it does await the
responses without freezing the main UI.

Changing sleep(0) to yield makes fib_hell run about twice as slow as fib,
which makes sense because it creates twice the amount of context switches.
It does freeze the UI though, so it's not useful. From what I understand,
SWI's wasm version will occasionally sleep for 100 ms or so to let the UI
do some work. I have not yet experimented with an approach like this.

I think a more robust solution would be to use web workers. AFAIK this is
what the Ciao playground does. This would be analogous to calling
pl_create() N times and running them in N different OS threads in the
native version. This is more or less how the Go version of Trealla works,
too. It should be possible to add native predicates that do this as
well, akin to SWI's engines.

Anyway, this is a bit off-topic. I just wanted to say that yes, you can
control the event loop. It boils down to orchestrating calls to pl_redo()
and using yield/0 to break out of the query loop.

**Comment by Jean-Luc-Picard-2021** [comment]
> Here's an example of how trealla-js specializes sleep/0 by calling into JS: https://github.com/guregu/trealla/blob/02e188c7420016c7f63829af3450225a61fcebbc/library/wasm_js.pl#L128-L141

Did you run this test case with your wasm.js? How fast is the 
sleep/1 there? I am assuming a non-skip sleep, where sleep(0) yields.
```
/* Trealla Prolog */
?- time(fib(29,_)).
   % Time elapsed 0.716444s
   true.
?- time(fib_hell(29,_)).
   % Time elapsed 63.235085s
   true.
```
For comparison, non WASM but pure JavaScript, no browser, 
but command line, using nodeJS setImmediate(), I get:
```
/* Dogelog Player */
?- time(fib(29,_)).
% Wall 1680 ms, gc 73 ms, 3466927 lips
true.
?- time(fib_hell(29,_)).
% Wall 4717 ms, gc 57 ms, 2293147 lips
true.
```
Pretty nice? Here is a summary. Somehow Scryer Prolog also
manages to call a fast sleep/1:

|System|fib/2|fib_hell/2|Ratio|
|---|---|---|---|
|Jekejeke|365|541|148%|
|Scryer|383|674|175%|
|Dogelog|1680|4717|281%|
|SWI|193|2823|1463%|
|Trealla|716|63235|8832%|

**Comment by guregu** [comment]
That is pretty nice! Here's what I measured:
I'm replacing sleep(0) with yield/0 for the WASM version as sleep(0) currently does the minimal 16ms or whatever browsers force on you. yield/0 does a context switch from wasm guest -> host so I think that's what we want to measure.
To illustrate the overhead of the JS calls, I tried running it on node and calling into setImmediate as well.

```prolog
mysleep :- js_eval("return new Promise((resolve => setImmediate(resolve)));", _).
```


```
?- time(fib(29, _)).
    % Time elapsed 0.603000s

% sleep(0) replaced with yield
?- time(fib_hell(29, _)).
    % Time elapsed 1.444000s

% sleep(0) replaced with mysleep (uses setImmediate, running in Node)
?- time(fib_hell_setimmediate(29, _)).
    % Time elapsed 14.889000s
```

And for a baseline, here's it from Dogeloge player browser version [from here](https://www.dogelog.ch/littab/doclet/en/docs/04_tutor/10_reference/example02/package.html):

```
?- time(fib(29, _)).
% Wall 1819 ms, gc 33 ms, 3202042 lips
```

So for trealla-js it's roughly about 2x overhead for yielding, and 20x overhead for calling into JS (ouch). But, I have a feeling like this isn't a fair measurement, because trealla-js doesn't have a heartbeat system like SWI so it yields and then immediately goes back and blocks the UI thread unless it has a promise to await. You can see that `sleep(10)` in trealla-js does not block the main thread while it's sleeping because it's wrapped in a promise. I'm pretty sure SWI is automatically inserting sleeps somewhere kind of like this. While yield/0 in trealla-js does switch contexts, I'm not sure if it ever actually gives other code a chance to run. In other words, these numbers are probably too fast. Need to investigate further... 🤔

**Comment by Jean-Luc-Picard-2021** [comment]
Do you have auto-yield? If you run `fib(29,_)` what you call **baseline**,
it will do auto-yield. To run it without auto-yield, you need to call
it like with `shield/1` wrapped. There is a very small difference usually,

since auto-yield is called around 60-times per second:
```
/* Dogelog Player, nodeJS auto-yield via setImmediate() */
?- time(fib(29,_)).
% Wall 1728 ms, gc 65 ms, 3370624 lips
true.

/* Dogelog Player, nodeJS no auto-yield */
?- shield(time(fib(29,_))).
% Wall 1698 ms, gc 74 ms, 3430107 lips
true.
```
But difference between browser and commandline nodeJS is quite dramatic,
because browser doesn't have setImmediate(), and setTimeout() has a documented 
capping to 4ms. I get this difference now, measuring the auto-yield

in the browser. Always calling fib/2 and not fib_hell/2:
```
/* Dogelog Player, browser auto-yield via setTimeout() */
?- time(fib(29,_)).
% Wall 2520 ms, gc 19 ms, 2311316 lips
true.

/* Dogelog Player, browser no auto-yield */
?- shield(time(fib(29,_))).
% Wall 1503 ms, gc 15 ms, 3875131 lips
true.
```
So browser Chrome V8 is faster than nodeJS, if I switch off auto-yield.

**Comment by guregu** [comment]
Ah, that makes sense. trealla-js doesn't have any kind of auto-yield at the moment. 

I tried running your example with shield/1 in my browser:
```
?- shield(time(fib(29, _))).
% Wall 931 ms, gc 14 ms, 6255984 lips
true.
```

This is much closer to what trealla-js gets, so it seems like it's only fair to compare them turning auto-yield off. I  think I need to add auto-yield to trealla-js :-)

~~Are you calling setTimeout/setImmediate after every redo? I'm not sure what the best way to actually yield is.~~ Nevermind, answered my question already. I see!

**Comment by guregu** [comment]
BTW, have you looked into using Web Workers to run it in a separate thread? I think with that maybe there would be no need for auto-yielding. https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API

**Comment by Jean-Luc-Picard-2021** [comment]
With auto-yield you can run two tasks in parallel, and they get interleaved. 
Currently in Dogelog Player, again only using fib/2 and not fib_hell/2:
```
task(N) :- fib(N, X), write('fib('), write(N), write(')='), write(X), nl.
```
The result will be shown almost simultaneously. The tasks will more or
less spend the same time in fib/2, without time measurement yet, but
one can see by eye that the two tasks show their result almost at the same time:
```
?- create_task(task(30)), create_task(task(30)), sleep(6000).
fib(30)=1346269
fib(30)=1346269
true.
```
Since no Worker is involved, the two tasks share the same knowledge
base, for example the predicate fib/2. No need to load it twice into two Workers.
And since they are coroutines, no locking or atomics is needed. You could 

also make an example with dynamic predicates. If you need some locking 
or atomics you can use `shield/1` instead, to mark a critical region. As already
demonstrate shield/1 does switch off auto-yield, but it has also other use cases,

than only speeding up computation. If you keep auto-yield on, it can be used
as a synchronization primitive. Instead `with_mutex(M, (G1, ..., Gn))`, you simply
do `shield((G1, .., Gn))`, and you have your critical region as well.

**Comment by guregu** [comment]
That's cool, thanks for explaining. Auto-yield by default with optional critical sections makes a lot of sense, I like it. I think it is more elegant than using mutexes for sure. I'll play around with adding something similar.
I wonder if there's some way we can get around the lack of setImmediate in browsers. The only thing I know of is this: https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame but I have a feeling like it's not the right thing for a query engine, haha.

**Comment by Jean-Luc-Picard-2021** [comment]
Usually if you have GUI and an end-user fiddling with the mouse
or some animation, the 4ms aren't a problem. It keeps your browser alive.
If you are lucky the graphics window is updated in this time slice.

**Comment by guregu** [comment]
I messed with trealla-js a bit and got it to actually yield (via setTimeout) from yield/0. Now the UI won't freeze anymore. https://github.com/guregu/trealla-js/pull/24

New times for fib_hell(29, _) from the browser:
```
  %% with yield/0:
  % Time elapsed 1.708000s

  %% no yield:
  % Time elapsed 0.607000s
```

So about a 3x slowdown in return for a responsive UI. Probably worth it.
I need to make Trealla automatically yield though, currently only works with manually calling yield/0.

**Comment by guregu** [comment]
@infradig any good spots to stick an auto-yield in the query loop? Like have the backtracking within pl_query/pl_redo automatically do a yield/0 every N milliseconds.

**Comment by infradig** [comment]
Could add a pl_yield call that you stick in the loop.

**Comment by guregu** [comment]
It's kind of tricky because we can't call `pl_yield` while a query is running in the browser, it hogs the main thread. So we need to give it a deadline before it starts evaluating.
Maybe add a deadline to pl_query? Something like this:

```c
// automatically yield in ~200ms
pl_query(pl, "repeat.", &subq, current_time() + 200);
do {
    if (query_did_yield(subq)) {
        // (do browser yielding stuff)
        
        // set new deadline
        pl_yield_at(subq, current_time() + 200);
        continue
    }
    // regular processing stuff
} while (pl_redo(subq))
```

**Comment by infradig** [comment]
Ok, added it in devel branch.

**Comment by guregu** [comment]
Maybe a dumb question but does the devel branch stuff interrupt the eval loop?
Yesterday I was messing with it a bit and got it to kind of work by inserting some checks in `start()` in `query.c`, but I'm pretty sure I was doing something wrong.
I'm thinking it should work similar to `g_tpl_interrupt` but checking the current time against the deadline, if that makes any sense. I think SWI implements theirs by checking the time every N backtracks.

**Comment by infradig** [comment]
I just added the pl_* calls you wanted.

**Comment by guregu** [comment]
Thanks. Reading the code it seems like `q->yield_at` isn't referenced by the eval loop, so I was thinking it wouldn't get interrupted. I'll play around with it.

**Comment by infradig** [comment]
I modelled it on the reference loop you showed. Definitely no interrupt.

**Comment by guregu** [comment]
Ah, I see. Sorry, my example was confusing.  Could we get it to interrupt the loop? I realize that could impact performance but we can #ifdef it out of the non-wasm version. Unfortunately I think this is the only way to get browsers to cooperate.

**Comment by infradig** [comment]
Interrupt the loop by dropping out? Like on N backtracks?

**Comment by infradig** [comment]
I could check the timeout in the internal backtrack loop and yield there.

**Comment by guregu** [comment]
> I could check the timeout in the internal backtrack loop and yield there.
> […](#)

Yep, that's exactly what I'm lookg for. Thanks!

**Comment by infradig** [comment]
Ok, check the code in devel now.

**Comment by guregu** [comment]
Awesome, looks great. I'll try integrating it tonight and report back.

**Comment by guregu** [comment]
I played around with it a bit, hacking in a msec parameter for pl_query so I can give it a time before it starts.
It's very close to working perfectly.
Would it be possible to get it to yield before `query->retry` is set?

In the `fib_hell/2` example, adding this to the top makes it autoyield nicely:
```prolog
fib_hell(_, _) :- fail.
```

But without the extra fail, it won't check the yield_at time until after fib_hell/2 finishes. 
(Sorry, I initially said on backtracking, but it looks like what we need is a check before any call.)

**Comment by infradig** [comment]
Can you make the changes that work for you and then send them in when done.
I can't work on this today.

**Comment by guregu** [comment]
Sounds good. I'll mess with it today and send a PR if I get it working. No rush, take a peek when you get some spare time.

**Comment by Jean-Luc-Picard-2021** [comment]
I am doing auto-yield at the call port. This allows auto-yielding
also for example such as:
```
?- repeat, fail.
```
Subsequently I can press Ctrl-C now in Dogelog Player,
and the above gets aborted, although it runs single threaded.

I am using the auto-yield frequency also as the frequency
how signals are polled. Auto-yielding and signal polling are

currently the same.

**Comment by infradig** [comment]
> Do you have an engine abstraction in Trealla Prolog? Like this here from SWI-Prolog:
> 
> > 11.3 Engine predicate reference
> > https://www.swi-prolog.org/pldoc/man?section=engine-predicates
> 
> So everybody can write an event loop in Prolog itself? Well the API is not perfect, a lot is missing right now.

Engines compatible with SWI-Prolog was added. They share the Database (like SWI) and so are not thread-safe. Consider it experimental at this stage. Also my understanding of all the calls may be imperfect. I think engine_yield/1 is wrong... it should yield/block until the term is read.

**Comment by Jean-Luc-Picard-2021** [comment]
engine_yield/1 blocks. You will not go beyond
engine_yield/1 if there is not a pairing engine_next/2.

**Comment by Jean-Luc-Picard-2021** [comment]
I opened a new issue. And closed this issue.

**Comment by Jean-Luc-Picard-2021** [comment]
In Dogelog Player the heartbeat is transparent, it amounts to
sleep(0) but executed via a yield. But not some yield that would
be visible to the end-user. In as far Dogelog Player has no engines

for the end-user, only callbacks and tasks:
```
/* Dogelog Player */
?- findall(X, (X=1,'$YIELD'(0);X=2), L).
L = [1, 2].
```

**Comment by infradig** [comment]
Leaving this open if Guregu has any questions.

**Comment by Jean-Luc-Picard-2021** [comment]
I have unsubscribed from this issue.

**Comment by guregu** [comment]
I think I got it working! Thanks for your help @Jean-Luc-Picard-2021 and Andrew.

Demo:

https://user-images.githubusercontent.com/131059/223721191-12b405c1-1c74-4131-bda5-4407959ccd97.mov

The background will freeze and stop changing colors if it doesn't yield, but we can see that it is smooth now and immediately updates the UI. Yay! (Yes, I've been staring at this psychedelic debug screen the last couple days :-))

One question for @infradig:
Is it OK to move the yield check to right after the goal is evaluated? Here: https://github.com/guregu/trealla/blob/ea2116e0e8c2f1534d9b3a2b8cfebed2ea01df34/src/query.c#L1755-L1763
Everything seems to work nicely. I added a very arbitrary "run it every 10 goals" condition which I believe SWI also does. Need to look up what their defaults are. Calling the time functions in wasm is a context switch so reducing them makes it about 4x faster.
I also tried doing it before the goal call, like how g_tpl_interrupt works, but that seemed to cause weird things to happen, maybe because the wasm toplevel is implemented in Prolog without any critical section type things.

**Comment by infradig** [comment]
I'd say if it seems to work you're on the right track. Great work! I'll
look at the code tomorrow.

**Comment by guregu** [comment]
Made a PR with the changes: #136. Feel free to change stuff or ask me to change stuff.
I borrowed SWI's default values for the yield check goal interval and time and now it runs about twice as fast and very smoothly. I'm pretty happy with the results, will enable this by default for trealla-js I think.

**Comment by guregu** [comment]
Just released v0.14.0 of trealla-js which includes these changes. Thanks everyone!
Feel free to close this now.
